### PR TITLE
`/ui2/cl_json`: add inherit functionality

### PR DESCRIPTION
**Pull request by oblomov-dev**
open-abap is not working when a class inherits /ui2/json:
https://github.com/abap2UI5/abap2UI5/actions/runs/7583976799/job/20656717865?pr=793

added:
* constructor
* is_compressable
* serialize_int (and added the code of the serialize method)
* added attributes used in the constructor

next step / issue: 
* adjusting serialize_int with respecting the value mv_extended.

I have no idea if this inheriting concept is easily implementable in open-abap? before i start making changes here without knowing how this all specifically works, maybe its better you comment on this first.

### CODE REVIEW

**Comment by larshp** [comment]
inheritance should work in open-abap, so all good

**Comment by larshp** [comment]
@oblomov-dev if you dont have vscode installed, you can try https://github.com/github/dev, it might be easier

**Comment by oblomov-dev** [comment]
yes github codespaces (i think it is the same) is my second home after eclipse :) i will check it!

**Comment by oblomov-dev** [comment]
ok, fixed all the linter issues now, all tests are green.

i skipped implementing the call of the "is_compressable" method, means from my understanding: you can inherit /ui2/cl_json now but when you create a unit test based on a change implemented in is_compressable, open-abap will give a different result than the normal abap stack but maybe for the first step, this is enough.

File: src/json/#ui2#cl_json.clas.abap [hunk 1]
```diff
@@ -21,6 +21,13 @@ CLASS /ui2/cl_json DEFINITION PUBLIC.
       CHANGING
         data             TYPE data.
 
+    METHODS is_compressable
```
Comment (larshp) [inline src/json/#ui2#cl_json.clas.abap:24]: move this method to PROTECTED SECTION ?

File: src/json/#ui2#cl_json.clas.abap [hunk 2]
```diff
@@ -38,6 +45,25 @@ CLASS /ui2/cl_json DEFINITION PUBLIC.
...
+        compress TYPE abap_bool DEFAULT abap_false
+        pretty_name TYPE pretty_name_mode DEFAULT pretty_mode-none
+        assoc_arrays TYPE abap_bool DEFAULT abap_false
+        ts_as_iso8601 TYPE abap_bool DEFAULT abap_false.
+
+    DATA mv_compress TYPE abap_bool.
+    DATA mv_pretty_name TYPE string.
+    DATA mv_assoc_arrays TYPE abap_bool.
+    DATA mv_ts_as_iso8601 TYPE abap_bool.
+    DATA mv_extended TYPE abap_bool.
```
Comment (larshp) [inline src/json/#ui2#cl_json.clas.abap:65]: also protected section?

File: src/json/#ui2#cl_json.clas.abap [hunk 3]
```diff
@@ -118,11 +141,11 @@ CLASS /ui2/cl_json IMPLEMENTATION.
...
           lv_index = sy-tabix.
-          r_json = r_json && serialize(
-            data          = <any>
-            pretty_name   = pretty_name
-            compress      = compress
-            ts_as_iso8601 = ts_as_iso8601 ).
+          r_json = r_json && serialize_int( data = <any> ).
+*            pretty_name   = mv_pretty_name
+*            compress      = mv_compress
+*            ts_as_iso8601 = mv_ts_as_iso8601
```
Comment (larshp) [inline src/json/#ui2#cl_json.clas.abap:147]:
```suggestion
```

File: src/json/#ui2#cl_json.clas.abap [hunk 4]
```diff
@@ -135,21 +158,20 @@ CLASS /ui2/cl_json IMPLEMENTATION.
...
           ENDIF.
-          r_json = r_json && serialize(
-            data          = <any>
-            pretty_name   = pretty_name
-            compress      = compress
-            ts_as_iso8601 = ts_as_iso8601 ).
+          r_json = r_json && serialize_int( data = <any> ).
+*           pretty_name   = pretty_name
+*           compress      = compress
+*           ts_as_iso8601 = ts_as_iso8601 ).
```
Comment (larshp) [inline src/json/#ui2#cl_json.clas.abap:174]:
```suggestion
```

File: src/json/#ui2#cl_json.clas.abap [hunk 5]
```diff
@@ -162,11 +184,10 @@ CLASS /ui2/cl_json IMPLEMENTATION.
...
         ASSIGN data->* TO <any>.
-        r_json = serialize(
-          data          = <any>
-          pretty_name   = pretty_name
-          compress      = compress
-          ts_as_iso8601 = ts_as_iso8601 ).
+        r_json = serialize_int( data = <any> ).
+*          pretty_name   = pretty_name
+*          compress      = compress
+*          ts_as_iso8601 = ts_as_iso8601 ).
```
Comment (larshp) [inline src/json/#ui2#cl_json.clas.abap:190]:
```suggestion
```

**Review by larshp — CHANGES_REQUESTED** [review]
see comments

**Comment by oblomov-dev** [comment]
see this review and comment function for the first time, nice! code is adjusted.

**Review by larshp — APPROVED** [review]
thanks 👍

**Comment by larshp** [comment]
if the compress inheritance and stuff doesnt work, suggest as next step adding a unit test for it in https://github.com/oblomov-dev/open-abap-core/blob/main/src/json/%23ui2%23cl_json.clas.testclasses.abap

**Comment by oblomov-dev** [comment]
@larshp 
ok, would there also be the chance to debug the code? can i for example rename it to zui2_cl_json, copy it to an abap stack and run the unit tests or what is the recommended way?

**Comment by larshp** [comment]
that might work

typically I dont debug much, when there is a simple unit test for it its easy to find the bug

standalone debuggin also works a bit,
image

**Comment by oblomov-dev** [comment]
ah ok interesting, i will try it out :)

**Pull request merged**